sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala

```diff
@@ -56,33 +58,89 @@ case class CreateArray(children: Seq[Expression]) extends Expression {
   }
 
   override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
-    val arrayClass = classOf[GenericArrayData].getName
-    val values = ctx.freshName("values")
-    ctx.addMutableState("Object[]", values, s"this.$values = null;")
+    val array = ctx.freshName("array")
 
-    ev.copy(code = s"""
-      this.$values = new Object[${children.size}];""" +
+    val et = dataType.elementType
+    val evals = children.map(e => e.genCode(ctx))
+    val isPrimitiveArray = ctx.isPrimitiveType(et)
+    val primitiveTypeName = if (isPrimitiveArray) ctx.primitiveTypeName(et) else ""
+    val (preprocess, arrayData, arrayWriter) =
+      GenArrayData.getCodeArrayData(ctx, et, children.size, isPrimitiveArray, array)
+
+    ev.copy(code =
```

Comment:
can you refactor it a little bit? The logic gets more complicated and it's hard to read when you put it in `ev.copy(...)`